app/models/concepts_graph.rb

```diff
@@ -0,0 +1,98 @@
+# ConceptGraph transforms a track configuration to a form compatible for manipulating for
+# displaying the progression of the track
+class ConceptsGraph
+  attr_reader(
+    :language,
+    :nodes,
+    :edges,
+    :node_lookup_by_concept,
+    :node_lookup_by_index,
+    :adjacency,
+    :has_cycle
+  )
+
+  def initialize(track_config)
+    @language = track_config['language']
+    @nodes = make_nodes(track_config['exercises'])
+    @node_lookup_by_concept = make_concept_lookup
+    @node_lookup_by_index = make_index_lookup
+    @edges = make_edges
+    @adjacency = GraphUtils.make_adjacency_list(self)
+    @has_cycle = GraphUtils.contains_cycle?(adjacency)
+  end
+
+  # Layout computes the ordering of the exercises in a track by creating
+  # a graph then traversing it to discover and memoize the progression
+  def to_layout
+    return [] if nodes.empty?
+    return :error if has_cycle
+
+    node_indegrees = GraphUtils.find_indegree(adjacency)
+    source_nodes = GraphUtils.find_sources(node_indegrees)
+    exercise_depths = GraphUtils.find_depths(source_nodes, adjacency)
+
+    blank_layout = Array.new(exercise_depths.max + 1) { [] }
+    exercise_depths.each_with_index.each_with_object(blank_layout) do |(depth, index), layers|
+      layers[depth] << graph.node_lookup_by_index[index]
+    end
+  end
+
+  private
+  def make_nodes(exercises)
+    exercises.each_with_index.reduce([]) do |a, (exercise, i)|
+      a << Node.new(
+        index: i,
+        slug: exercise['slug'],
+        uuid: exercise['uuid'],
+        concepts: exercise['concepts'],
+        prerequisites: exercise['prerequisites']
+      )
+    end.freeze
+  end
+
+  def make_concept_lookup
```

Comment:
Rather than setting the instance variable, let's just pop this in a memoized method.

```suggestion
  memoize
  def node_lookup_by_concept
```